### ISSUE TITLE
Introduction of Manipulative Param Converters

### ISSUE THREAD

**Issue by maennchen** [FriendsOfSymfony/FOSRestBundle#1595]
I'm one of the authors of https://github.com/ibrows/rest-bundle which is an extension of the Fos Rest Bundle.

We use a construct called Manipulative Param Converters. There is for example a `PatchParamConverter` which applies a JSON Patch (RFC 6902) onto an object loaded by another param converter.

Example:
```php
/**
 * @param Comment $comment
 * @ParamConverter(
 *     name="comment",
 *     converter="patch",
 *     class="AppBundle\Entity\Comment",
 *     options={
 *         "source" = "doctrine.orm"
 *     }
 * )
 *
 * @Patch("/{comment}")
 */
public function patchAction(Comment $comment)
{
    // ...
}
```

This will:
 * Load the Object via the param converter named in `options[source]`
 * Apply the specified converter onto that object

Is there interest to refactor & integrate this into the FosRestBundle?

**Comment by Ener-Getick** [comment]
I think it would be interesting to have such param converter.
But [it looks like](https://github.com/sensiolabs/SensioFrameworkExtraBundle/pull/436) param converter are meant to die, so shouldn't we wait for their alternative?

Besides that, would you be interested to port some of your bundle features into fosrestbundle? I looked quickly its doc and your EtagListener looks especially interesting to me.

**Comment by iltar** [comment]
To me this looks like a bad approach (sorry for bursting in, came here through the linked PR). This means that adding a specific annotation would cause manipulation of objects you receive, while this manipulation should be initiated in your controller. As it is, you are altering the state of managed objects at a point you have no control over. That means that if an event listener triggers a flush for any reason at all, you have no idea if this will flush your entity, not to mention that you can't guarantee a valid state of your entity at all.

Just my 2 cents.

**Comment by maennchen** [comment]
@iltar Yes, but I don't see a point in repeating myself in every controller for a very defined task (like applying a JSON Patch onto an object.)

This way an API Controller can be built very fast and is tested thoroughly.

Also EventListeners shouldn't just flush random objects. That is not a problem of a ParamConverter.

**Comment by maennchen** [comment]
@Ener-Getick Yes sure, I'll make separate PR's for the listeners that make sense. (Soon, I don't have time right now.)

I don't really like the approach of this Arg Annotation. If you just call one simple function, it's easy. If you have a more complex converter (like setting the serializer context group) for the body converter, this will get very messy.

**Comment by iltar** [comment]
@maennchen parameter converters are meant to transform an incoming parameter into an object, not to execute logic.

**Comment by maennchen** [comment]
@iltar I have no doubt about that they are not meant to execute code, but lacking a better instrument we implemented it like that.

Even though it may be not the perfect solution, it gets the job done without repeating the same code in every controller.

If you have a better Idea on how to solve a generic approach to execute JSON Patches, Link / Unlink requests etc. on incoming parameters, I'd love to change the code to a better suited implementation.

**Comment by iltar** [comment]
```php
public function fooAction(Comment $comment)
{
    $this->someCommandBus->handle(new PathEntity($comment));
}

// or service style
public function fooAction(Comment $comment)
{
    $this->patcher->patch($comment);
}
```

Just because it's convenient, doesn't make it a good solution (especially not this one).
